File: frontend/services/listings.ts
import type { AxiosInstance } from "axios";

import { apiClient } from "@lib/api-client";
import type { SearchProperty } from "@services/search";

export type ListingProperty = SearchProperty;

export type ListingDetailResponse = {
  property: ListingProperty;
  images: string[];
};

export type ListingSubmissionPayload = {
  property_type: string;
  listing_type: string;
  title: string;
  description?: string | null;
  address?: string | null;
  city: string;
  state: string;
  size_sqft?: number | null;
  price?: number | null;
  clear_height?: number | null;
  loading_docks?: number | null;
  contact_name: string;
  contact_email: string;
};

export type ListingSubmissionResult = { id: string; status: string };

export type FeaturedListingsResponse = {
  listings: ListingDetailResponse[];
};

export type AgentProfileResponse = {
  name: string;
  email?: string | null;
  phone?: string | null;
  properties: ListingProperty[];
};

export type ListingSubmissionStatus = "pending" | "approved" | "rejected";

export type ListingSubmissionItem = {
  id: string;
  status: ListingSubmissionStatus;
  property_type?: string | null;
  listing_type?: string | null;
  title?: string | null;
  city?: string | null;
  state?: string | null;
  size_sqft?: number | null;
  price?: number | null;
  contact_name?: string | null;
  contact_email?: string | null;
  created_at?: string | null;
};

export class ListingsService {
  constructor(private readonly http: AxiosInstance) {}

  async getFeaturedListings(
    options?: { signal?: AbortSignal },
  ): Promise<FeaturedListingsResponse> {
    const { data } = await this.http.get<FeaturedListingsResponse>("/listings/featured", {
      signal: options?.signal,
    });
    return data;
  }

  async getListing(
    propertyId: string,
    options?: { signal?: AbortSignal },
  ): Promise<ListingDetailResponse> {
    const { data } = await this.http.get<ListingDetailResponse>(`/listings/${propertyId}`, {
      signal: options?.signal,
    });
    return data;
  }

  async submitListing(
    payload: ListingSubmissionPayload,
    options?: { signal?: AbortSignal },
  ): Promise<ListingSubmissionResult> {
    const { data } = await this.http.post<ListingSubmissionResult>(
      "/listing-submissions",
      payload,
      { signal: options?.signal },
    );
    return data;
  }

  async getAgent(
    broker: string,
    options?: { signal?: AbortSignal },
  ): Promise<AgentProfileResponse> {
    const { data } = await this.http.get<AgentProfileResponse>(
      `/agents/${encodeURIComponent(broker)}`,
      { signal: options?.signal },
    );
    return data;
  }

  async listSubmissions(options?: { signal?: AbortSignal }): Promise<ListingSubmissionItem[]> {
    const { data } = await this.http.get<ListingSubmissionItem[]>("/listing-submissions", {
      signal: options?.signal,
    });
    return data;
  }

  async updateSubmission(
    id: string,
    status: ListingSubmissionStatus,
    options?: { signal?: AbortSignal },
  ): Promise<ListingSubmissionItem> {
    const { data } = await this.http.patch<ListingSubmissionItem>(
      `/listing-submissions/${encodeURIComponent(id)}`,
      { status },
      { signal: options?.signal },
    );
    return data;
  }
}

export const listingsService = new ListingsService(apiClient);

File: backend/app/schemas/listings.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.properties import Properties


class ListingDetailResponse(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    property: Properties
    images: list[str]


class ListingSubmissionCreate(BaseModel):
    """A property listing submitted via the public 'list your property' form."""

    model_config = ConfigDict(str_strip_whitespace=True)

    property_type: str = Field(min_length=1, max_length=120)
    listing_type: str = Field(min_length=1, max_length=40)
    title: str = Field(min_length=1, max_length=300)
    description: str | None = Field(default=None, max_length=5000)
    address: str | None = Field(default=None, max_length=500)
    city: str = Field(min_length=1, max_length=120)
    state: str = Field(min_length=1, max_length=120)
    size_sqft: float | None = Field(default=None, ge=0)
    price: float | None = Field(default=None, ge=0)
    clear_height: float | None = Field(default=None, ge=0)
    loading_docks: int | None = Field(default=None, ge=0)
    contact_name: str = Field(min_length=1, max_length=200)
    contact_email: EmailStr


class ListingSubmissionResponse(BaseModel):
    id: str
    status: str


class ListingSubmissionItem(BaseModel):
    """Admin view of a submitted listing (extra DB columns ignored)."""

    model_config = ConfigDict(extra="ignore")

    id: str
    status: str
    property_type: str | None = None
    listing_type: str | None = None
    title: str | None = None
    city: str | None = None
    state: str | None = None
    size_sqft: float | None = None
    price: float | None = None
    contact_name: str | None = None
    contact_email: str | None = None
    created_at: str | None = None


class ListingSubmissionStatusUpdate(BaseModel):
    status: Literal["pending", "approved", "rejected"]


class AgentProfileResponse(BaseModel):
    """A listing broker/agent and the properties they have listed."""

    name: str
    email: str | None = None
    phone: str | None = None
    properties: list[Properties]
